### ISSUE TITLE
admin: data race (config pull at start)

### ISSUE THREAD

**Issue by dtelyukh** [caddyserver/caddy#4260]
### 1a. Operating system and version

```
Manjaro 21.1.0 Pahvo (x86_64 Linux 5.4.135-1-MANJARO)
```

### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
2.4.3
```

### 1c. Go version (if building Caddy from source; run `go version`)

```
1.16.6
```

## 2. Description

data race found 

Caddy config:
```
{
	"admin": {
		"config": {
			"load": {
				"module": "http",
				"url": "http://localhost:8080/config"
			}
		}
	}
}

```

Config from http://localhost:8080/config:

```
https://guix.localhost {
    tls internal

    reverse_proxy {
        header_up host guix.gnu.org
        to https://guix.gnu.org
    }

    handle {
        @assets {
            path_regexp \.(ico|css|js|gif|jpg|jpeg|png|svg|woff|woff2|otf)($|\?)
        }
        header @assets {
            Cache-Control "public, max-age=31536000"
        }
    }
}
```

<details><summary>Data race at start</summary>

```
==================
WARNING: DATA RACE
Write at 0x0000049288c8 by goroutine 43:
  github.com/caddyserver/caddy/v2.replaceLocalAdminServer()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/admin.go:332 +0x5ea
  github.com/caddyserver/caddy/v2.run()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:379 +0x695
  github.com/caddyserver/caddy/v2.unsyncedDecodeAndRun()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:276 +0x171
  github.com/caddyserver/caddy/v2.finishSettingUp.func1()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:492 +0x234

Previous read at 0x0000049288c8 by goroutine 35:
  github.com/caddyserver/caddy/v2.replaceLocalAdminServer.func2()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/admin.go:343 +0x54

Goroutine 43 (running) created at:
  github.com/caddyserver/caddy/v2.finishSettingUp()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:489 +0x793
  github.com/caddyserver/caddy/v2.run()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:457 +0x5ab
  github.com/caddyserver/caddy/v2.unsyncedDecodeAndRun()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:276 +0x171
  github.com/caddyserver/caddy/v2.changeConfig()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:168 +0x604
  github.com/caddyserver/caddy/v2.Load()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:114 +0xf3
  github.com/caddyserver/caddy/v2/cmd.cmdRun()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/commandfuncs.go:214 +0x4f5
  github.com/caddyserver/caddy/v2/cmd.Main()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/main.go:85 +0x2f4
  main.main()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/caddy/main.go:41 +0x2f

Goroutine 35 (running) created at:
  github.com/caddyserver/caddy/v2.replaceLocalAdminServer()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/admin.go:342 +0x669
  github.com/caddyserver/caddy/v2.run()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:379 +0x695
  github.com/caddyserver/caddy/v2.unsyncedDecodeAndRun()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:276 +0x171
  github.com/caddyserver/caddy/v2.changeConfig()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:168 +0x604
  github.com/caddyserver/caddy/v2.Load()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/caddy.go:114 +0xf3
  github.com/caddyserver/caddy/v2/cmd.cmdRun()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/commandfuncs.go:214 +0x4f5
  github.com/caddyserver/caddy/v2/cmd.Main()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/main.go:85 +0x2f4
  main.main()
      /home/dionysius/Документы/Исходники/go/gitlab.com/activecdn/caddy/cmd/caddy/main.go:41 +0x2f
==================
```

</details>

**Comment by mholt** [comment]
Thanks for the report. Looks like we need a mutex around `localAdminServer`.

**Comment by stangles** [comment]
I took a shot at a fix for this in #4274.
